### PR TITLE
Try to handle lombok.Builder without crashing.

### PR DESCRIPTION
**Pull request by lazaroclapp**
IMPORTANT:
This PR doesn't solve the fundamental incompatibilities between
lombok's code transformation approach and Error Prone's assumption
that the AST in memory corresponds to the original code for error
reporting.

What this change does is prevent NullAway for reporting on
`@Builder` generated code in most cases, while not completely
turning off checking at the top-level class scope.

We accomplish this by hard-coding the following two defaults
into NullAway:
* First, we make `@Builder` a default external initialization
  annotation for the fields of the class in which it appears.
* Second, we make `@lombok.Generated` an excluded
  class annotation, skipping analysis for classes marked
  as generated.

The use of the later class requires that lombok be configured
with `addLombokGeneratedAnnotation`. We add instructions
in the `README` to this effect.

We include tests and a new sample project running lombok
to more adequately test the interaction between NullAway and
the Lombok annotation processor for this and future changes.

### CODE REVIEW

File: test-java-lib-lombok/src/main/java/com/uber/lombok/LombokBuilderInit.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+import javax.annotation.Nullable;
+import lombok.Builder;
+
+@Builder
+public class LombokBuilderInit {
+  private String field;
+  @Builder.Default private String fieldWithDefault = "Default";
+  @Nullable private String nullableField;
+}
```
Comment (lazaroclapp) [inline test-java-lib-lombok/src/main/java/com/uber/lombok/LombokBuilderInit.java:33]: For reference, this becomes:

```
package com.uber.lombok;

import javax.annotation.Nullable;
import lombok.Builder;

@Builder()
public class LombokBuilderInit {
    private String field;
    @lombok.Builder.Default()
    private String fieldWithDefault;
    @Nullable()
    private String nullableField;
    
    @java.lang.SuppressWarnings(value = "all")
    private static String $default$fieldWithDefault() {
        return "Default";
    }
    
    @java.lang.SuppressWarnings(value = "all")
    LombokBuilderInit(final String field, final String fieldWithDefault, @Nullable()
    final String nullableField) {
        super();
        this.field = field;
        this.fieldWithDefault = fieldWithDefault;
        this.nullableField = nullableField;
    }
    
    @java.lang.SuppressWarnings(value = "all")
    public static class LombokBuilderInitBuilder {
        @java.lang.SuppressWarnings(value = "all")
        private String field;
        @java.lang.SuppressWarnings(value = "all")
        private boolean fieldWithDefault$set;
        @java.lang.SuppressWarnings(value = "all")
        private String fieldWithDefault$value;
        @java.lang.SuppressWarnings(value = "all")
        private String nullableField;
        
        @java.lang.SuppressWarnings(value = "all")
        LombokBuilderInitBuilder() {
            super();
        }
        
        @java.lang.SuppressWarnings(value = "all")
        public LombokBuilderInit.LombokBuilderInitBuilder field(final String field) {
            this.field = field;
            return this;
        }
        
        @java.lang.SuppressWarnings(value = "all")
        public LombokBuilderInit.LombokBuilderInitBuilder fieldWithDefault(final String fieldWithDefault) {
            this.fieldWithDefault$value = fieldWithDefault;
            fieldWithDefault$set = true;
            return this;
        }
        
        @java.lang.SuppressWarnings(value = "all")
        public LombokBuilderInit.LombokBuilderInitBuilder nullableField(@Nullable()
        final String nullableField) {
            this.nullableField = nullableField;
            return this;
        }
        
        @java.lang.SuppressWarnings(value = "all")
        public LombokBuilderInit build() {
            String fieldWithDefault$value = this.fieldWithDefault$value;
            if (!this.fieldWithDefault$set) fieldWithDefault$value = LombokBuilderInit.$default$fieldWithDefault();
            return new LombokBuilderInit(this.field, fieldWithDefault$value, this.nullableField);
        }
        
        @java.lang.Override()
        @java.lang.SuppressWarnings(value = "all")
        public java.lang.String toString() {
            return "LombokBuilderInit.LombokBuilderInitBuilder(field=" + this.field + ", fieldWithDefault$value=" + this.fieldWithDefault$value + ", nullableField=" + this.nullableField + ")";
        }
    }
    
    @java.lang.SuppressWarnings(value = "all")
    public static LombokBuilderInit.LombokBuilderInitBuilder builder() {
        return new LombokBuilderInit.LombokBuilderInitBuilder();
    }
}
```

before Error Prone gets to see it. As of 2.4.0, EP does not acknowledge `@SuppressWarnings("all")`.

**Comment by lazaroclapp** [comment]
This is more of a proposal to start addressing long-standing issue #321 , which pops up again and again. It's a bit hacky, and probably will need some performance testing, and it certainly still doesn't fix the general issue of reporting errors inside EP on ASTs that were transformed by lombok, but it's the best I could think to do...

**Comment by msridhar** [comment]
Just FYI, there is a Lombok setting `lombok.addLombokGeneratedAnnotation` (documented on the [configuration page](https://projectlombok.org/features/configuration)) that makes Lombok add `@Generated` annotations on code it generates.  Maybe it could help in identifying when you are inside a generated Lombok Builder more efficiently?  Not sure, just thought I'd mention it.

**Comment by lazaroclapp** [comment]
Updated PR description

File: test-java-lib-lombok/build.gradle [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+tasks.withType(JavaCompile) {
+  if (!name.toLowerCase().contains("test")) {
+    options.errorprone {
+      check("NullAway", CheckSeverity.ERROR)
+      check("UnusedVariable", CheckSeverity.OFF) // We are not the only checker that fails on Lombok
+      option("NullAway:AnnotatedPackages", "com.uber")
+      option("NullAway:UnannotatedSubPackages", "com.uber.lib.unannotated")
+      option("NullAway:TryHandleLombok", "true")
```
Comment (msridhar) [inline test-java-lib-lombok/build.gradle:43]: Remove this now?

File: nullaway/build.gradle [hunk 3]
```diff
@@ -60,6 +60,7 @@ dependencies {
     testCompile deps.test.commonsLang
     testCompile deps.test.commonsLang3
     testCompile project(":test-library-models")
+    testCompile deps.test.lombok
```
Comment (msridhar) [inline nullaway/build.gradle:63]: So I guess Lombok will run on all our test inputs?  I guess this shouldn't cause a problem as long as they don't use Lombok.  But maybe it's sufficient to just compile `test-java-lib-lombok` and not also have a Lombok unit test?

Comment (lazaroclapp) [reply nullaway/build.gradle:63]: Lombok will not run on any tests, this merely adds the library dependency, we are not adding the annotation processor itself. So, the unit test exclusively tests the defaults for `@Builder` but nothing about `@lombok.Generated` or the processor itself.

**Pull request merged**